cpp/ct2.cpp

```diff
@@ -241,3 +247,129 @@ extern "C" lean_obj_res ct2_encode(b_lean_obj_arg _input_tokens) {
 
   return arr;
 }
+
+
+extern "C" uint8_t init_premise_embeddings(b_lean_obj_arg matrix_path) {
+  const char *mat_path = lean_string_cstr(matrix_path);
+  if (!exists(mat_path)) {
+    return false;
+  }
+  if (premise_embeddings != nullptr) {
+    delete premise_embeddings;
+  }
+  
+  auto d = npy::read_npy<double>(mat_path);
+  std::vector<double> data = d.data;
+  std::vector<unsigned long> shape = d.shape;
+  bool fortran_order = d.fortran_order;
+
+  std::vector<float> data_f;
+  data_f.resize(data.size());
+  std::transform(data.begin(), data.end(), data_f.begin(),
+                 [](double d) { return static_cast<float>(d); });
+
+  std::vector<int64_t> shape_i64;
+  shape_i64.resize(shape.size());
+  std::transform(shape.begin(), shape.end(), shape_i64.begin(),
+                 [](unsigned long ul) { return static_cast<int64_t>(ul); });
+
+  premise_embeddings = new ctranslate2::StorageView(shape_i64, data_f, ctranslate2::Device::CPU);
```

Comment:
The device shouldn't be hardcoded.

Reply:
Just saw that we hardcoded device to be CPU when we were initializing encoder before. Would we like to change this as well?

Reply:
where?

Reply:
On line 218 (in the latest version) in `init_ct2_encoder`.

Reply:
I see. Can you change it to something similar to what we're doing for the generator?

Reply:
Sure!

Reply:
Done :)